Userland/Libraries/LibCrypto/BigFraction/BigFraction.cpp

```diff
@@ -0,0 +1,252 @@
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+BigFraction BigFraction::operator-(BigFraction const& rhs) const
+{
+    return *this + (-rhs);
+}
+
+BigFraction BigFraction::operator*(BigFraction const& rhs) const
+{
+    auto tmp = *this;
+    tmp.m_numerator.set_to(tmp.m_numerator.multiplied_by(rhs.m_numerator));
+    tmp.m_denominator.set_to(tmp.m_denominator.multiplied_by(rhs.m_denominator));
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+BigFraction BigFraction::operator-() const
+{
+    return { m_numerator.negated_value(), m_denominator };
+}
+
+BigFraction BigFraction::invert() const
+{
+    return BigFraction { 1 } / *this;
+}
+
+BigFraction BigFraction::operator/(BigFraction const& rhs) const
+{
+    VERIFY(rhs.m_numerator != "0"_bigint);
+
+    auto tmp = *this;
+    tmp.m_numerator.set_to(m_numerator.multiplied_by(rhs.m_denominator));
+    tmp.m_denominator.set_to(m_denominator.multiplied_by(rhs.m_numerator.unsigned_value()));
+
+    if (rhs.m_numerator.is_negative())
+        tmp.m_numerator.negate();
+
+    tmp.reduce();
+
+    return tmp;
+}
+
+bool BigFraction::operator<(BigFraction const& rhs) const
+{
+    return (*this - rhs).m_numerator.is_negative();
+}
+
+bool BigFraction::operator==(BigFraction const& rhs) const
+{
+    return m_numerator == rhs.m_numerator && m_denominator == rhs.m_denominator;
+}
+
+BigFraction::BigFraction(double d)
+{
+    bool negative = false;
+    if (d < 0) {
+        negative = true;
+        d = -d;
+    }
+    i8 current_pow = 0;
+    while (AK::pow(10.0, (double)current_pow) <= d)
+        current_pow += 1;
+    current_pow -= 1;
+    double epsilon = 1e-6;
+    unsigned decimal_places = 0;
+    while (d >= epsilon || current_pow >= 0) {
+        m_numerator.set_to(m_numerator.multiplied_by(Crypto::SignedBigInteger { 10 }));
+        i8 digit = (u64)(d * AK::pow(0.1, (double)current_pow)) % 10;
+        m_numerator.set_to(m_numerator.plus(Crypto::UnsignedBigInteger::create_from(digit)));
+        d -= digit * AK::pow(10.0, (double)current_pow);
+        if (current_pow < 0) {
+            ++decimal_places;
+            m_denominator.set_to(Crypto::NumberTheory::Power("10"_bigint, Crypto::UnsignedBigInteger::create_from(decimal_places)));
+        }
+        current_pow -= 1;
+        if (decimal_places > 6)
+            break;
+    }
+    m_numerator.set_to(negative ? (m_numerator.negated_value()) : m_numerator);
+}
+
+double BigFraction::to_double() const
+{
+    // FIXME: very naive implementation
+    return m_numerator.to_double() / m_denominator.to_double();
+}
+
+void BigFraction::set_to_0()
+{
+    m_numerator.set_to_0();
+    m_denominator.set_to(1);
+}
+
+BigFraction BigFraction::rounded(unsigned rounding_threshold) const
+{
+    auto const get_last_digit = [](auto const& integer) {
+        return integer.divided_by("10"_bigint).remainder;
+    };
+
+    auto const res = m_numerator.divided_by(m_denominator);
+    BigFraction tmp { res.quotient };
```

Comment:
Please move the temporary values, bigints are not free to copy.
(many instances)

Reply:
I'm not totally sure of what I've done, tell me if it's not good

Reply:
Looks alright~